### PR TITLE
Add diagm for vectors

### PR DESCRIPTION
**Pull request by eulerkochy**
This in reference to JuliaLang/LinearAlgebra.jl#609 . I've added the missing function definition.

### CODE REVIEW

File: base/dict.jl [hunk 1]
```diff
@@ -223,7 +223,7 @@ function rehash!(h::Dict{K,V}, newsz = length(h.keys)) where V where K
     h.count = count
     h.ndel = 0
     h.maxprobe = maxprobe
-    @assert h.age == age0
+    h.age == age0 || error("versions of the dict do not match")
```
Comment (StefanKarpinski) [inline base/dict.jl]: You've left this change in from the previous PR.

Comment (StefanKarpinski) [inline base/dict.jl]: You've left this change in from the previous PR.

**Review by StefanKarpinski — CHANGES_REQUESTED** [review]
Thanks for the contribution! This should also add a test for the method, checking that it works as expected as well as documentation and a NEWS entry.

File: stdlib/LinearAlgebra/src/dense.jl [hunk 2]
```diff
@@ -291,6 +291,10 @@ function diagm_container(kv::Pair{<:Integer,<:BitVector}...)
     return falses(n, n)
 end
 
+# Defining diagm() for an AbstractVector
```
[the remark this answers is not in the document]
Comment (eulerkochy) [reply stdlib/LinearAlgebra/src/dense.jl]: Ok, my bad!

File: base/dict.jl (hunk 1 above)
Comment (eulerkochy) [reply base/dict.jl]: Well, what to do about it?

Comment (StefanKarpinski) [reply base/dict.jl]: Get rid of that change on your branch.

**Comment by eulerkochy** [comment]
I've removed the un-necessary commit from the branch!

File: stdlib/LinearAlgebra/src/dense.jl [hunk 3]
```diff
@@ -291,6 +291,10 @@ function diagm_container(kv::Pair{<:Integer,<:BitVector}...)
...
+
+diagm(v::AbstractVector) = diagm(0=>v)
```
Comment (StefanKarpinski) [inline stdlib/LinearAlgebra/src/dense.jl]: This got unfixed.

File: base/dict.jl (hunk 1 above)
Comment (StefanKarpinski) [inline base/dict.jl]: It's still there.

File: stdlib/LinearAlgebra/src/dense.jl (hunk 3 above)
Comment (eulerkochy) [reply stdlib/LinearAlgebra/src/dense.jl]: How the hell?

File: base/dict.jl (hunk 1 above)
Comment (eulerkochy) [reply base/dict.jl]: :(

File: stdlib/LinearAlgebra/src/dense.jl (hunk 3 above)
Comment (eulerkochy) [reply stdlib/LinearAlgebra/src/dense.jl]: I force-pushed the latest commit!

Comment (eulerkochy) [reply stdlib/LinearAlgebra/src/dense.jl]: Any alternative?

File: base/dict.jl (hunk 1 above)
Comment (eulerkochy) [reply base/dict.jl]: Should I create a new branch and send another PR?

File: stdlib/LinearAlgebra/src/dense.jl (hunk 3 above)
Comment (StefanKarpinski) [reply stdlib/LinearAlgebra/src/dense.jl]: You made a remote commit through the GitHub UI. The force push overwrote it. You'd want to pull the change and do an interactive rebase to drop the unwanted commit.

Comment (eulerkochy) [reply stdlib/LinearAlgebra/src/dense.jl]: Ok!

**Comment by StefanKarpinski** [comment]
At this point I would just do `git fetch` to pull the latest master and then `git reset origin/master` and edit the code until the diff looks the way you want it to. This is a small change and the git gymnastics required to do this via commits and rebasing aren't really worth it.

**Comment by eulerkochy** [comment]
Is it fine now?

File: stdlib/LinearAlgebra/src/dense.jl (hunk 3 above)
[the remark this answers is not in the document]
Comment (KristofferC) [reply stdlib/LinearAlgebra/src/dense.jl]: This was not fixed.

**Comment by KristofferC** [comment]
> Is it fine now?

If you look at https://github.com/JuliaLang/julia/pull/31125/files you can see that many irrelevant files seems to have been deleted in this PR. There is also still no test added.

**Comment by eulerkochy** [comment]
Tests as in documented examples of the function call? If so, what the format to be followed?

**Comment by KristofferC** [comment]
Tests as in testing the functionality of the new feature (using the `@test` macro), see e.g. https://github.com/JuliaLang/julia/blob/master/CONTRIBUTING.md#writing-tests.

Also we could update all the instances of syntax like ` diagm(0 => D.diag)` in the code to use `diagm(D.diag)` instead.

**Comment by eulerkochy** [comment]
Ok! But can I make a separate PR for that? This PR has too clumsy for a lapse on my part and I'm proud of this fact. :(

**Comment by KristofferC** [comment]
Sure, that can very well be in another PR.

File: stdlib/LinearAlgebra/src/dense.jl [hunk 4]
```diff
@@ -291,6 +291,7 @@ function diagm_container(kv::Pair{<:Integer,<:BitVector}...)
     return falses(n, n)
 end
 
+diagm(v::AbstractVector) = diagm( 0 => v)
```
Comment (KristofferC) [inline stdlib/LinearAlgebra/src/dense.jl]:
```suggestion
diagm(v::AbstractVector) = diagm(0 => v)
```

**Comment by eulerkochy** [comment]
Ok, what's left to do in this PR? Can anyone tell? @KristofferC @StefanKarpinski

**Comment by pkofod** [comment]
> Ok, what's left to do in this PR? Can anyone tell? @KristofferC @StefanKarpinski

No need to ping when people are already in the conversation :)

I did note the conversation above, but would it really be such a hassle to simply include the tests in this PR? It's easier to find at a later point if it's in the same PR, and you'll actually  be 110% confident that your changes work as intended. Even the smallest changes can have unintended consequences, so it's just good practice IMO. I get the eagerness to have the PR merged, but honestly, there's not going to be a new release of Julia tomorrow (I don't think there is at least ! ) so if this is merged now or if it's merged three days from now WITH tests really doesn't matter from an end-users perspective - so I'd say the latter (wait and merge with tests) makes sense. Even for such a "simple" (though appreciated) PR.

**Comment by eulerkochy** [comment]
Thanks for the advice, and well I was just curious to know what remaining work was left so that I could complete it. Seems like to add tests for it!

**Comment by eulerkochy** [comment]
Just trying to reopen!

File: stdlib/LinearAlgebra/src/dense.jl [hunk 5]
```diff
@@ -291,6 +291,22 @@ function diagm_container(kv::Pair{<:Integer,<:BitVector}...)
...
+# Examples
+```jldoctest
+julia> diagm([1,2,3])
+3×3 Array{Int64,2}:
+ 1  0  0
+ 0  2  0
+ 0  0  3
+```
+"""
+
```
Comment (StefanKarpinski) [inline stdlib/LinearAlgebra/src/dense.jl]:
```suggestion
```

File: stdlib/LinearAlgebra/test/dense.jl [hunk 6]
```diff
@@ -84,6 +84,15 @@ bimg  = randn(n,2)/2
         end
     end # for eltyb
 
+@testset "Test diagm for vectors" begin
+    @test diagm(zeros(50)) == diagm(0 => zeros(50))
+    @test diagm(ones(50)) == diagm(0 => ones(50))
+    for i = 1:50
```
Comment (StefanKarpinski) [inline stdlib/LinearAlgebra/test/dense.jl]: I don't really think this needs to be done 50 times. Once is enough to test that it's working.

**Comment by StefanKarpinski** [comment]
Getting there!

File: stdlib/LinearAlgebra/test/dense.jl (hunk 6 above)
Comment (eulerkochy) [reply stdlib/LinearAlgebra/test/dense.jl]: Ok, so I scrape it?

Comment (StefanKarpinski) [reply stdlib/LinearAlgebra/test/dense.jl]: Yes, just remove the loop and keep the inner part.

File: NEWS.md [hunk 7]
```diff
@@ -52,6 +52,7 @@ Standard library changes
 * Eigenvalues λ of general matrices are now sorted lexicographically by (Re λ, Im λ) ([#21598]).
 * `one` for structured matrices (`Diagonal`, `Bidiagonal`, `Tridiagonal`, `Symtridiagonal`) now preserves
   structure and type. ([#29777])
+* `diagm` can now construct a diagonal matrix from a vector (#[31125]).
```
Comment (StefanKarpinski) [inline NEWS.md]:
```suggestion
* `diagm(v)` is now a shorthand for `diagm(0 => v)` (#[31125]).
```
`diagm` has always been able to construct a diagonal matrix from a vector, it's just a question of how to spell it.

**Review by StefanKarpinski — APPROVED** [review]
Looking good. Glad you stuck it out!

**Review by fredrikekre — APPROVED** [review]

**Pull request merged**